x-pack/filebeat/input/entityanalytics/provider/okta/internal/okta/okta_test.go

```diff
@@ -148,85 +149,166 @@ func Test(t *testing.T) {
-	)
-	var wantUsers []User
-	err := json.Unmarshal([]byte(msg), &wantUsers)
-	if err != nil {
-		t.Fatalf("failed to unmarshal user data: %v", err)
-	}
 
-	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		u, err := url.Parse(r.RequestURI)
+	t.Run("device", func(t *testing.T) {
+		query := make(url.Values)
+		query.Set("limit", "200")
+		devices, _, err := GetDeviceDetails(context.Background(), http.DefaultClient, host, key, "", query, limiter, window)
 		if err != nil {
-			t.Errorf("unexpected error parsing request URI: %v", err)
+			t.Fatalf("unexpected error: %v", err)
 		}
-		if u.Path != "/api/v1/users" {
-			t.Errorf("unexpected API endpoint: got:%s want:%s", u.Path, "/api/v1/users")
+		if *logResponses {
+			b, err := json.Marshal(devices)
+			if err != nil {
+				t.Errorf("failed to marshal devices for logging: %v", err)
+			}
+			t.Logf("devices: %s", b)
 		}
-		if got := r.Header.Get("accept"); got != "application/json" {
-			t.Errorf("unexpected Accept header: got:%s want:%s", got, "application/json")
+		for _, d := range devices {
+			users, _, err := GetDeviceUsers(context.Background(), http.DefaultClient, host, key, d.ID, query, OmitCredentials, limiter, window)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			b, err := json.Marshal(users)
+			if err != nil {
+				t.Errorf("failed to marshal users for logging: %v", err)
+			}
+			t.Logf("users: %s", b)
 		}
-		if got := r.Header.Get("authorization"); got != "SSWS "+key {
-			t.Errorf("unexpected Authorization header: got:%s want:%s", got, "SSWS "+key)
+	})
+}
+
+var localTests = []struct {
+	name   string
+	msg    string
+	id     string
+	fn     func(ctx context.Context, cli *http.Client, host, key, user string, query url.Values, lim *rate.Limiter, window time.Duration) (any, http.Header, error)
+	mkWant func(string) (any, error)
+}{
+	{
+		// Test case constructed from API-returned value with details anonymised.
+		name: "users",
+		msg:  `[{"id":"userid","status":"STATUS","created":"2023-05-14T13:37:20.000Z","activated":null,"statusChanged":"2023-05-15T01:50:30.000Z","lastLogin":"2023-05-15T01:59:20.000Z","lastUpdated":"2023-05-15T01:50:32.000Z","passwordChanged":"2023-05-15T01:50:32.000Z","type":{"id":"typeid"},"profile":{"firstName":"name","lastName":"surname","mobilePhone":null,"secondEmail":null,"login":"name.surname@example.com","email":"name.surname@example.com"},"credentials":{"password":{"value":"secret"},"emails":[{"value":"name.surname@example.com","status":"VERIFIED","type":"PRIMARY"}],"provider":{"type":"OKTA","name":"OKTA"}},"_links":{"self":{"href":"https://localhost/api/v1/users/userid"}}}]`,
```

Comment:
would it make sense to move these messages to a file of their own and assign them to constants? I'm thinking if more tests are added in future the test file will look messy and be hard to navigate.

Reply:
I favour locality in testing. So I would rather not.